File: ArraysAndHashing/Easy/121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.kt
class Solution {
    fun maxProfit(prices: IntArray): Int {
        // buying stock should be the min
        // selling stock should be the max
        // iterate through array keeping track of selling stock
        // if an element is greater than buying stock keep track of difference
        var buyingStock = prices[0]
        var profit = 0
        prices.forEach { p -> 
            if (p < buyingStock) buyingStock = p
            else {
                profit = max(profit, p - buyingStock)
            }
        }

        return profit
    }
}

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        currentMin = float('inf')
        currentMax = float('-inf')
        profit = 0

        for n in prices:
            if n < currentMin:
                currentMin = n # 0
                currentMax = currentMin
            elif n > currentMax:
                currentMax = n # 2
                print(f"current max is {currentMax}")
                profit = max(profit, currentMax - currentMin)
            # print(f"current max is {currentMax} and min is {currentMin} with n being {n}")
        return profit